### PR TITLE
Introduce TrackData

### PR DESCRIPTION
**Pull request by dimitry-**
Move number of timers, min, max and timer_chains
to TrackData.

Test: Run orbit instrument some functions and capture

### CODE REVIEW

File: src/OrbitGl/GraphTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/GraphTrack.cpp:107]: maybe add /\*depth=\*/

File: src/OrbitGl/PageFaultsTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/PageFaultsTrack.cpp:178]: maybe add /\*depth=\*/

File: src/OrbitGl/ThreadTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/ThreadTrack.cpp:445]: maybe add /\*depth=\*/

File: src/OrbitGl/ThreadTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/ThreadTrack.cpp:459]: While we are at it, I'd replace auto* by TimerChain*

File: src/OrbitGl/TimeGraph.h
Comment (pierricgimmig) [inline src/OrbitGl/TimeGraph.h:108]: It's unfortunate that this is not const anymore. It seems to me like this shouldn't modify the TimeGraph in any way. What was the reason behind removing const? Same for the functions below.

File: src/OrbitGl/TimerTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/TimerTrack.cpp:274]: I know it was there before, but I'm not a huge fan of auto :-) Consider using TimerChain instead.

File: src/OrbitGl/TimerTrack.cpp
Comment (pierricgimmig) [inline src/OrbitGl/TimerTrack.cpp:395]: nit. same thing here for auto...

**Review by pierricgimmig — APPROVED** [review]
Hi Dimitry, thanks for this. Looks good in general, but please see minor comments.

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [reply src/ClientData/include/ClientData/TrackData.h:20]: long term we want to get rid of TimerInfo and split it into smaller pieces. Currently the structure preserves the way it was stored before. Let's revisit it after the TextBox is removed. @karupayun is working on this.

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [reply src/ClientData/include/ClientData/TrackData.h:69]: insert and emplace take a pair (map::value) and return exactly the same result. insert_or_assing is more readable. I do not think there is any advantages in using insert/emplace, or am I missing something?

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [inline src/ClientData/include/ClientData/TrackData.h:39]: we cannot return const TimerChain here because there are parts of code that modify returned TextBox. And this is a reason why iterator cannot be made const. I do not like it either but before the code was basically lying about being const where it was not really const. I decided to make it explicit. My hope is that once we remove TextBox we can restore constness here properly - by making timer chain iterator const as well, like it was done for BlockChain

File: src/OrbitGl/TimeGraph.h
Comment (dimitry-) [reply src/OrbitGl/TimeGraph.h:108]: I explained it above - in TrackData::GetChains() method

**Review by dimitry- — COMMENTED** [review]
I am currently finishing with tests. will address auto* and other commend after that, in the meantime I answered questions about const/insert_or_assing and unique_ptr

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [reply src/ClientData/include/ClientData/TrackData.h:21]: done

File: src/OrbitGl/GraphTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/GraphTrack.cpp:107]: done.

File: src/OrbitGl/PageFaultsTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/PageFaultsTrack.cpp:178]: done.

File: src/OrbitGl/ThreadTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/ThreadTrack.cpp:445]: done.

File: src/OrbitGl/ThreadTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/ThreadTrack.cpp:459]: done.

File: src/OrbitGl/TimerTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/TimerTrack.cpp:274]: done.

File: src/OrbitGl/TimerTrack.cpp
Comment (dimitry-) [reply src/OrbitGl/TimerTrack.cpp:395]: done.

File: src/ClientData/include/ClientData/TrackData.h
Comment (karupayun) [inline src/ClientData/include/ClientData/TrackData.h:51]: Novelty for me, I didn't know about this function. :)

File: src/OrbitGl/Track.cpp
Comment (karupayun) [inline src/OrbitGl/Track.cpp:26]: better moving to the constuctor instead than in the .h file? Why?

File: src/OrbitGl/Track.h
Comment (karupayun) [inline src/OrbitGl/Track.h:65]: Why you are erasing this? This was part of the PR to not being able to move the SchedulerTrack.

File: src/ClientData/include/ClientData/TrackData.h
Comment (karupayun) [inline src/ClientData/include/ClientData/TrackData.h:14]: I understand that this is the current state of the code, but in the future, Track only will have a bunch of Panes which therefore will have the Events (Timers/Samples), do we agree on that?

File: src/ClientData/TrackDataTest.cpp
Comment (karupayun) [inline src/ClientData/TrackDataTest.cpp:40]: This is not a real case, this timer shouldn't be in depth == 1.

What do you think instead in adding also an additional timer from 8 to 11 at depth 0? Therefore, this could be a real case and you also can test if it is possible to have 2 timers at depth 0.

Nit: You also can split and make a test for an individual timer and another for this case.

**Review by karupayun — COMMENTED** [review]
Nice PR! I've left a few comments, will approve the PR after that.

File: src/ClientData/TrackDataTest.cpp
Comment (dimitry-) [reply src/ClientData/TrackDataTest.cpp:40]: it is technically possible - this tests that no matter which depth timer is added to it updates min and max.

File: src/OrbitGl/Track.h
Comment (dimitry-) [reply src/OrbitGl/Track.h:65]: it is currently unused - was it supposed to be used in the future PRs?

File: src/OrbitGl/Track.cpp
Comment (dimitry-) [reply src/OrbitGl/Track.cpp:26]: well, I was following clang-tidy on this one. :)

File: src/OrbitGl/Track.h
Comment (karupayun) [reply src/OrbitGl/Track.h:65]: @pierricgimmig are you still planning to use it?

File: src/ClientData/TrackDataTest.cpp
Comment (karupayun) [reply src/ClientData/TrackDataTest.cpp:40]: Yes, I know, but I would prefer modify it as I said, because it will be also a valid case

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [reply src/ClientData/include/ClientData/TrackData.h:39]: done.

File: src/ClientData/TrackDataTest.cpp
Comment (dimitry-) [reply src/ClientData/TrackDataTest.cpp:40]: done.

File: src/ClientData/include/ClientData/TrackData.h
Comment (dimitry-) [reply src/ClientData/include/ClientData/TrackData.h:14]: This structure is not going to be used by CaptureDataProvider, this is just a container for Legacy TimerInfo stuff.

File: src/ClientData/include/ClientData/TrackData.h
Comment (pierricgimmig) [reply src/ClientData/include/ClientData/TrackData.h:71]: Ah ok that's fair enough, thanks for the clarification.

File: src/OrbitGl/TimeGraph.h
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.h:108]: Could we add a `[[nodiscard]] *const* TimerChain* GetChain(uint64_t depth) *const*` method?

Comment (dimitry-) [reply src/OrbitGl/TimeGraph.h:108]: yep - tried that, the TimerChainItartor needs to be converted to const as well otherwise we won't be able to iterate over it. And this is blocked on TextBox removal. At least it is easier to do that way. We cold introduce const and non-const iterators, but I think all iterators over chains should be const.

File: src/OrbitGl/Track.h
Comment (pierricgimmig) [reply src/OrbitGl/Track.h:65]: The PR introduced a new method, IsFilterable that basically does what Movable does, so it's fine to remove it.

File: src/OrbitGl/TimeGraph.h
Comment (dimitry-) [reply src/OrbitGl/TimeGraph.h:108]: As discussed offline this leads to a lot of duplication I am going to revisit it after @karupayun change.

**Review by karupayun — COMMENTED** [review]
Thanks for addressing all my comments! :)

**Pull request merged**
